lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -13,548 +13,1055 @@
+      emitCircuitError()
+          << "contains a 'companion' with id '"
+          << a.first.cast<IntegerAttr>().getValue().getZExtValue()
+          << "', but does not contain a GrandCentral 'parent' with the "
+             "same "
+             "id";
       return signalPassFailure();
-    AnnotationSet annotations(&op);
-
-    annotations.removeAnnotations([&](auto anno) {
-      // Insert an instantiated interface.
-      if (auto viewAnnotation = annotations.getAnnotation(
-              "sifive.enterprise.grandcentral.ViewAnnotation")) {
-
-        auto tpe = viewAnnotation.getAs<StringAttr>("type");
-        if (tpe && tpe.getValue() == "parent") {
-          auto name = viewAnnotation.getAs<StringAttr>("name");
-          auto defName = viewAnnotation.getAs<StringAttr>("defName");
-          auto guard = OpBuilder::InsertionGuard(builder);
-          builder.setInsertionPointToEnd(cast<FModuleOp>(op).getBodyBlock());
-          auto instance = builder.create<sv::InterfaceInstanceOp>(
-              circuitOp->getLoc(),
-              interfaces.lookup(defName.getValue()).getInterfaceType(), name,
-              builder.getStringAttr(
-                  "__" + op.getAttrOfType<StringAttr>("sym_name").getValue() +
-                  "_" + defName.getValue() + "__"));
-          instance->setAttr("doNotPrint", builder.getBoolAttr(true));
-          builder.setInsertionPointToStart(
-              op.getParentOfType<ModuleOp>().getBody());
-          auto bind = builder.create<sv::BindInterfaceOp>(
-              circuitOp->getLoc(), SymbolRefAttr::get(instance.sym_nameAttr()));
-          bind->setAttr(
-              "output_file",
-              hw::OutputFileAttr::get(
-                  builder.getStringAttr(""),
-                  builder.getStringAttr("bindings.sv"),
-                  /*exclude_from_filelist=*/builder.getBoolAttr(true),
-                  /*exclude_replicated_ops=*/builder.getBoolAttr(true),
-                  bind.getContext()));
-        }
-        return true;
-      }
-      // All other annotations pass through unmodified.
-      return false;
-    });
+    }
+  }
 
-    annotations.applyToOperation(&op);
+  // Check that a companion exists for every parent.
+  for (auto a : parentIDMap) {
+    if (companionIDMap.count(a.first) == 0) {
+      emitCircuitError()
+          << "contains a 'parent' with id '"
+          << a.first.cast<IntegerAttr>().getValue().getZExtValue()
+          << "', but does not contain a GrandCentral 'companion' "
+             "with the same id";
+      return signalPassFailure();
+    }
   }
 
-  // Populate interfaces.
-  for (auto &a : interfaceKeys) {
-    auto defName = a.first;
-    auto name = a.second;
+  LLVM_DEBUG({
+    // Print out the companion map, parent map, and all leaf values that
+    // were discovered.  Sort these by their keys before printing to make
+    // this easier to read.
+    SmallVector<IntegerAttr> ids;
+    auto sort = [&ids]() {
```

Comment:
lambdas have really made std::bind useless.

Reply:
Interesting... I had no idea that C++ had partial application... 🤔